library/Freckle/App/Database.hs

```diff
@@ -60,6 +74,32 @@ import UnliftIO.Exception (displayException)
 import UnliftIO.IORef
 import Yesod.Core.Types (HandlerData (..), RunHandlerEnv (..))
 
+-- | A monadic context in which a SQL backend is available
+--   for running database queries
+class Monad m => MonadSqlBackend m where
+  getSqlBackendM :: m SqlBackend
+
+instance (HasSqlBackend r, Monad m) => MonadSqlBackend (ReaderT r m) where
+  getSqlBackendM = asks getSqlBackend
+
+-- | Can be used e.g. to generalized from 'SqlPersistT' to 'MonadSqlBackend'
+liftSql :: MonadSqlBackend m => ReaderT SqlBackend m a -> m a
+liftSql (ReaderT f) = getSqlBackendM >>= f
+
+-- | The constraint @'MonadSqlTx' db m@ indicates that @m@ is a monadic
+--   context that can run @db@ actions within a SQL transaction.
+--   Typically this means that @m@ has access to a connection pool and
+--   @db@ has access to a connection.
+class (MonadSqlBackend db, Monad m) => MonadSqlTx db m | m -> db where
+  -- | Runs the action in a SQL transaction
+  runSqlTx :: HasCallStack => db a -> m a
+
+class HasSqlBackend a where
+  getSqlBackend :: a -> SqlBackend
+
+instance HasSqlBackend SqlBackend where
+  getSqlBackend = id
```

Comment:
My spidey sense is tingling a bit that we have/need this but also `HasSqlPool`. Seems to be very overlapping concerns (get a connection so we can `runDB`).

Assuming there's a valid reason for both, should we include a mirroring `HasSqlBackend (HandlerData child site)` too?

Reply:
They seem very different to me, being able to get a pool versus being able to get a single connection.

I don't see what the purpose of `HasSqlBackend (HandlerData child site)` would be. Won't the `site` always have a pool, not a single connection?

Reply:
Yeah, I guess they are practically different and I was conflating them because to me it's conceptually the same with the Pool being an implementation details.

```hs
runDB :: (MonadUnliftIO m, MonadReader env m, HasSqlPool env) => ...
runDB b = do
  p <- view sqlPoolL
  Pool.with p $ \b ->
    Persist.actuallyRunSqlQuery b q

runDB :: (MonadUnliftIO m, MonadReader env m, HasSqlBackend env) => ...
runDB b = do
  p <- getSqlBackendM
  Persist.actuallyRunSqlQuery b q
```

You can even imagine

```hs
class (MonadReader env m, HasSqlPool env) => MonadSqlBackend m where
  getSqlBackendM = do
    p <- view sqlPool
    Pool.with p pure
```

I mentioned spidey sense in my comment because I didn't have an actual concrete question or concern, these ideas were just floating around and thought there might be something there.

I think I've set to see us actually define a `HasSqlBackend` instead for anything other than `ReaderT SqlBackend` and I've yet to see us actually use `getSqlBackendM` anywhere -- so I'm still struggling to bring this into the realm of the practical.